core-bundle/src/Routing/ResponseContext/CoreResponseContextFactory.php

```diff
@@ -85,6 +90,26 @@ public function createContaoWebpageResponseContext(PageModel $pageModel): Respon
             $htmlHeadBag->setMetaRobots($pageModel->robots);
         }
 
+        if ($pageModel->enableCanonical && $pageModel->canonicalLink) {
+            /** @var Controller $controller */
+            $controller = $this->contaoFramework->getAdapter(Controller::class);
+            $url = $controller->replaceInsertTags($pageModel->canonicalLink, false);
+
+            // Ensure absolute links
+            // FIXME: Remove once we remove support for relative urls
+            if (!preg_match('#^https?://#', $url)) {
+                /** @var Environment $environment */
+                $environment = $this->contaoFramework->getAdapter(Environment::class);
+                $url = $environment->get('base').$url;
```

Comment:
This should be possible without the `Environment` class. I can try to make it work if you want.

Reply:
Feel free.

Reply:
See https://github.com/Toflar/contao/pull/7